geniza/corpus/management/commands/escr_alto_to_annotation.py

```diff
@@ -0,0 +1,249 @@
+
+# parse eScriptorium ALTO files, generate transcription IIIF annotations,
+# and attach to correct documents
+
+
+class AltoObject(xmlmap.XmlObject):
+    # all XmlObjects should inherit this class to get the root namespace
+    ROOT_NAMESPACES = {"alto": "http://www.loc.gov/standards/alto/ns-v4#"}
+
+
+class AltoPolygonalObject(AltoObject):
+    polygon = xmlmap.NodeField("alto:Shape/alto:Polygon/@POINTS", AltoObject)
+
+
+class Line(AltoPolygonalObject):
+    content = xmlmap.StringField("alto:String/@CONTENT")
+
+
+class TextBlock(AltoPolygonalObject):
+    id = xmlmap.StringField("./@ID")
+    lines = xmlmap.NodeListField("alto:TextLine", Line)
+    block_type_id = xmlmap.StringField("./@TAGREFS")
+
+
+class PrintSpace(AltoObject):
+    textblocks = xmlmap.NodeListField("alto:TextBlock", TextBlock)
+
+
+class EscriptoriumAlto(AltoObject):
+    filename = xmlmap.StringField(
+        "//alto:alto/alto:Description/alto:sourceImageInformation/alto:fileName"
+    )
+    printspace = xmlmap.NodeField(
+        "//alto:alto/alto:Layout/alto:Page/alto:PrintSpace", PrintSpace
+    )
+    tags = xmlmap.NodeListField("//alto:alto/alto:Tags/alto:OtherTag", AltoObject)
+
+
+class Command(BaseCommand):
+    # regex pattern for image filenames
+    filename_pattern = r"PGPID_(?P<pgpid>\d+)_(?P<shelfmark>[\w\-]+)_(?P<img>\d)\..+"
+
+    def add_arguments(self, parser):
+        # needs xml filenames as input
+        parser.add_argument(
+            "alto", metavar="ALTOXML", nargs="+", help="ALTO files to be processed"
+        )
+
+    def handle(self, *args, **options):
+        self.script_user = User.objects.get(username=settings.SCRIPT_USERNAME)
+
+        for xmlfile in options["alto"]:
+            self.stdout.write("Processing %s" % xmlfile)
+            self.ingest_xml(xmlfile)
+
+    def ingest_xml(self, xmlfile):
+        alto = xmlmap.load_xmlobject_from_file(xmlfile, EscriptoriumAlto)
+        # associate filename with pgpid
+        m = re.match(self.filename_pattern, alto.filename)
+        pgpid = int(m.group("pgpid"))
+        try:
+            doc = Document.objects.get(Q(id=pgpid) | Q(old_pgpids__contains=[pgpid]))
+        except Document.DoesNotExist:
+            self.stdout.write("Could not match document; skipping")
+            return
+
+        # we should be able to match the shelfmark portion to a manifest short_id
+        manifest = self.get_manifest(doc, m.group("shelfmark"))
+
+        # use canvas short_id = img number in sequence
+        img_number = int(m.group("img")) + 1
+        canvas = self.get_canvas(manifest, img_number)
+
+        if canvas:
+            canvas_uri = canvas.uri
+        else:
+            # create a placeholder canvas URI that contains textblock pk and canvas number
+            canvas_base_uri = "%siiif/" % doc.permalink
+            b = doc.textblock_set.first()
+            canvas_uri = f"{canvas_base_uri}textblock/{b.pk}/canvas/{img_number}/"
```

Comment:
That is actually the same logic we used to generate placeholder canvases previously. It's `TextBlock` as in the table relating `Fragment` and `Document`, so it doesn't correspond to a single block-level annotation, but rather a single fragment; it represents the structure Document > Fragment > Canvas. (This was because we were creating placeholder canvases with `img_number`s 1 and 2 for each imageless fragment, and needed a way to differentiate them.)

https://github.com/Princeton-CDH/geniza/blob/96ee37276dae240b30bd2d188059d4fcf8dc2135/geniza/corpus/views.py#L812-L819